enforcer-rules/src/main/java/org/apache/maven/plugins/enforcer/BanTransitiveDependencies.java

```diff
@@ -127,55 +130,27 @@ private static boolean searchTree(DependencyNode node, int level, ArtifactMatche
 
     @Override
     public void execute(EnforcerRuleHelper helper) throws EnforcerRuleException {
-        this.helper = helper;
-
-        if (excludes == null) {
-            excludes = Collections.emptyList();
-        }
-        if (includes == null) {
-            includes = Collections.emptyList();
-        }
-
-        final ArtifactMatcher exclusions = new ArtifactMatcher(excludes, includes);
-
-        DependencyNode rootNode = null;
-
+        MavenSession session;
         try {
-            MavenProject project = (MavenProject) helper.evaluate("${project}");
-            MavenSession session = (MavenSession) helper.evaluate("${session}");
-
-            ProjectBuildingRequest buildingRequest =
-                    new DefaultProjectBuildingRequest(session.getProjectBuildingRequest());
-            buildingRequest.setProject(project);
-
-            rootNode = createDependencyGraphBuilder().buildDependencyGraph(buildingRequest, null);
-        } catch (Exception e) {
-            throw new EnforcerRuleException("Error: Could not construct dependency tree.", e);
-        }
-
-        String message = getMessage();
-        StringBuilder generatedMessage = null;
-        if (message == null) {
-            generatedMessage = new StringBuilder();
+            session = (MavenSession) helper.evaluate("${session}");
+        } catch (ExpressionEvaluationException e) {
+            throw new RuntimeException(e);
         }
-
+        ArtifactTypeRegistry artifactTypeRegistry =
+                session.getRepositorySession().getArtifactTypeRegistry();
+        ArtifactMatcher exclusions = new ArtifactMatcher(excludes, includes);
+        Set<Dependency> directDependencies = session.getCurrentProject().getDependencies().parallelStream()
+                .map(d -> RepositoryUtils.toDependency(d, artifactTypeRegistry))
+                .collect(Collectors.toSet());
```

Comment:
parallelStream 😄 
I see other 7 places - are all thread safe?

Reply:
Collectors are not. noneMatch, anyMatch should be safe, but I can replace those too just to be on the safe side.